### PR TITLE
[libbeat] Add proxy_host support to kibana setup utilities

### LINKED ISSUE #8276: Add support for `proxy_url` to the dashboard setup task
The HTTP client used for accessing the Kibana API doesn't allow the user to specify a `proxy_url` option and doesn't support the `PROXY_HOST` environment variable, we should make it consistent with the Elasticsearch output and add support for it.

### PR DESCRIPTION
**Pull request by fearful-symmetry**
## What does this PR do?

This PR adds a `proxy_host` flag to the kibana setup code. See this issue: https://github.com/elastic/beats/issues/8276. This code is a tad complicated and I worry that there's a chunk of code I've forgotten somewhere. @urso does it look like I'm missing anything?

I still need to manually test it. Also, I'm not quite sure if I'm updating the correct `.yml.tmpl` file?

## Why is it important?

Other connections that go on within metricbeat adhere to `proxy_host` so this should too.

## Checklist
- [X] My code follows the style guidelines of this project
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## How to test this PR locally
 Still need to figure out how to test this myself. Presumably you can just..set up a squid proxy and then point `setup.kibana.proxy_host` at it?

## Related issues

- Closes elastic/beats#8276

### CODE REVIEW

File: libbeat/_meta/config.yml.tmpl [hunk 1]
```diff
@@ -39,6 +39,12 @@ setup.kibana:
   # IPv6 addresses should always be defined as: https://[2001:db8::1]:5601
   #host: "localhost:5601"
 
+  # Proxy Host
+  # Specify a proxy host that will route kibana setup tasks.
+  # This can be either an HTTP or HTTPS URL. If no value is set, 
+  # the HTTPS_PROXY and HTTP_PROXY environment variables will be checked.
+  #proxy_url: "https://proxy:3128"
```
Comment (urso) [inline libbeat/_meta/config.yml.tmpl:46]: nit (naming consistency): the setting is `proxy_url`, but in code and at the top of this comment we use `ProxyHost`. To be consistent with other HTTP clients we should use `proxy_url` always.

File: libbeat/kibana/client.go [hunk 2]
```diff
@@ -132,13 +135,26 @@ func NewClientWithConfig(config *ClientConfig) (*Client, error) {
...
+	var proxy func(*http.Request) (*url.URL, error)
+	if config.ProxyHost != "" {
+		proxyURL, err := url.Parse(config.ProxyHost)
+		if err != nil {
+			return nil, errors.Wrapf(err, "could not convert %s to url", config.ProxyHost)
+		}
+		proxy = http.ProxyURL(proxyURL)
+	} else if envCfg := httpproxy.FromEnvironment(); envCfg.HTTPProxy != "" || envCfg.HTTPSProxy != "" {
+		proxy = http.ProxyFromEnvironment
+	}
```
Comment (urso) [inline libbeat/kibana/client.go:148]: The httpproxy package looks interesting, as it also takes the NOPROXY environment variable into account. How about relying on `httpproxy` always. E.g.

```
proxyEnv := httpproxy.FromEnvironment()
if config.ProxyURL != nil {
  proxyEnv.HTTPProxy = config.ProxyURL
  proxyEnv.HTTPSProxy = config.ProxyURL
}
proxy := proxyEnv.ProxyFunc()
```

**Comment by fearful-symmetry** [comment]
Added a proxy disable setting. I _still_ can't figure out how to update the config reference docs. Am I even editing the correct file?

File: libbeat/kibana/client.go [hunk 3]
```diff
@@ -132,13 +136,28 @@ func NewClientWithConfig(config *ClientConfig) (*Client, error) {
...
+	proxyEnv := httpproxy.FromEnvironment()
+	if config.ProxyHost != "" {
+		proxyEnv.HTTPSProxy = config.ProxyHost
+		proxyEnv.HTTPProxy = config.ProxyHost
+	}
+	proxy := func(req *http.Request) (*url.URL, error) {
+		if config.ProxyDisable {
+			return nil, nil
+		}
+		return proxyEnv.ProxyFunc()(req.URL)
```
Comment (urso) [inline libbeat/kibana/client.go:150]: nit: `ProxyFunc` is a 'factory'. We should not call it over and over again. we can change init to read like this (consider to introduce a function configuring the 'Proxy'):

```
var proxy func(_ *http.Request) (*url.URL, error)
if !config.ProxyDisable {
  proxyEnv := ...
  urlProxy := proxyEnv.ProxyFunc()
  proxy = func(req *http.Request) (*url.URL, error) { return urlProxy(req.URL) }
}
```

**Comment by urso** [comment]
> I still can't figure out how to update the config reference docs. Am I even editing the correct file?

`make update` should generate the reference configuration files per beat.

The asciidoc documentation itself is not generated from the config.yml.tmpl file, but must be edited separately.

**Comment by urso** [comment]
Change LGTM.

Can we add more tests to see if these settings are effective (connect to fake HTTP server via fake proxy)?
At least this PR should get the `test-plan` label.

**Comment by fearful-symmetry** [comment]
@urso Yah, still need to do do that. Haven't fiddled much with proxy servers, so...not sure how.

**Comment by urso** [comment]
we've got some tests with proxy setup here: https://github.com/elastic/beats/blob/master/libbeat/outputs/elasticsearch/client_proxy_test.go

**Comment by urso** [comment]
@fearful-symmetry What is the status of this PR? Do we already support the proxy URL?

**Comment by fearful-symmetry** [comment]
@urso This is one of those things that got dropped once Integrations work picked up. Not sure what the current status is, and I'm currently full with stuff for 7.13.

**Comment by urso** [comment]
Closing. PR stalled and unmaintained

**Pull request closed without merging**